File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/ProviderCostEstimate.java
package com.foundgine.core.semantic.planning;

import java.util.Objects;

/** Provider-specific advisory estimate for executing a candidate semantic plan. */
public record ProviderCostEstimate(String provider,double estimatedExecutionCost,double estimatedRows,double confidence,CostEstimateProvenance provenance) {
    public ProviderCostEstimate(String provider,double estimatedExecutionCost){this(provider,estimatedExecutionCost,0d,.5d,null);}
    public ProviderCostEstimate { validateProvider(provider);validateNonNegative(estimatedExecutionCost,"estimatedExecutionCost");validateNonNegative(estimatedRows,"estimatedRows");if(Double.isNaN(confidence)||Double.isInfinite(confidence)||confidence<0||confidence>1)throw new IllegalArgumentException("Confidence must be between 0 and 1."); }
    public CostEstimateProvenance effectiveProvenance(){return provenance==null?CostEstimateProvenance.heuristic():provenance;}
    public static ProviderCostEstimate from(String provider,double cost){return new ProviderCostEstimate(provider,cost,0d,.5d,null);}
    public static ProviderCostEstimate from(String provider,double cost,double rows,double confidence,CostEstimateProvenance provenance){return new ProviderCostEstimate(provider,cost,rows,confidence,provenance);}
    private static void validateProvider(String provider){if(provider==null||provider.isBlank())throw new IllegalArgumentException("Provider is required.");}
    private static void validateNonNegative(double value,String name){if(Double.isNaN(value)||Double.isInfinite(value)||value<0)throw new IllegalArgumentException(name+" must be finite and non-negative.");}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/ProviderCostEstimate.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Provider-specific estimate for executing a candidate semantic plan.
/// The estimate is advisory: it may influence selection, but it cannot weaken
/// semantic equivalence or security requirements.
/// </summary>
public readonly record struct ProviderCostEstimate(
    string Provider,
    double EstimatedExecutionCost,
    double EstimatedRows = 0d,
    double Confidence = 0.5d,
    CostEstimateProvenance? Provenance = null)
{
    public CostEstimateProvenance EffectiveProvenance => Provenance ?? CostEstimateProvenance.Heuristic();

    public static ProviderCostEstimate From(
        string provider,
        double estimatedExecutionCost,
        double estimatedRows = 0d,
        double confidence = 0.5d,
        CostEstimateProvenance? provenance = null)
    {
        if (string.IsNullOrWhiteSpace(provider)) throw new ArgumentException("Provider is required.", nameof(provider));
        ValidateFiniteNonNegative(estimatedExecutionCost, nameof(estimatedExecutionCost));
        ValidateFiniteNonNegative(estimatedRows, nameof(estimatedRows));
        if (double.IsNaN(confidence) || double.IsInfinity(confidence) || confidence < 0d || confidence > 1d)
            throw new ArgumentOutOfRangeException(nameof(confidence));
        return new ProviderCostEstimate(provider, estimatedExecutionCost, estimatedRows, confidence, provenance);
    }

    private static void ValidateFiniteNonNegative(double value, string name)
    {
        if (double.IsNaN(value) || double.IsInfinity(value) || value < 0d)
            throw new ArgumentOutOfRangeException(name);
    }
}